torch/csrc/distributed/c10d/NanCheck.cu

```diff
@@ -11,13 +11,134 @@ namespace c10d {
+// EltPerPack = 16 / 2 = 8
+
+template <typename T>
+struct CheckBytePack<T, /*EltPerPack*/8> {
+  static __device__ __forceinline__ void check(BytePack* tmp) {
+    T* data = (T*)tmp;
+    if (isnan(data[0]) || isnan(data[1]) || isnan(data[2]) || isnan(data[3]) ||
+        isnan(data[4]) || isnan(data[5]) || isnan(data[6]) || isnan(data[7])) {
+          __trap();
+    }
+  }
+};
+
+//// End of templated functions for checking NaNs inside a BytePack
+
+
+// Fast-path check routine:
+// each thread will load and check 8 BytePacks in this routine
+
+// Create a tmp buffer of size 8, also unroll for loop by 8
+#define UNROLL 8
+
+template <typename T>
+__device__ __forceinline__ void checkChunk(BytePack* ptr) {
+  BytePack tmp[UNROLL];
+  int nWorkers = blockDim.x * gridDim.x;
+  // First load values from global memory into tmp buffer
+  #pragma unroll 8
+  for (int j = 0; j < UNROLL; j++) {
+    tmp[j] = ptr[nWorkers * j];
+  }
+  // Then check each BytePack in the tmp buffer
+  #pragma unroll 8
+  for (int j = 0; j < UNROLL; j++) {
+    CheckBytePack<T, sizeof(BytePack)/sizeof(T)>::check(tmp + j);
+  }
+  // Note: we separate the check from the load for efficient loading
+}
+
+// Align address of `ptr` up, to the alignment of `T`
+#define ALIGN_UP(ptr, T) (((uintptr_t)ptr + sizeof(T) - 1) / sizeof(T) * sizeof(T))
+
+// This is the host-facing kernel
+
 template <typename T>
 __global__ void checkForNaN(T* data, size_t size) {
-  size_t tid = blockIdx.x * blockDim.x + threadIdx.x;
-  size_t stride = blockDim.x * gridDim.x;
+  // Offset of current thread
+  size_t offset = blockIdx.x * blockDim.x + threadIdx.x;
+
+  // Align input address up to BytePack in case it is not
+  T* ptrAlign = (T*)ALIGN_UP(data, BytePack);
+  // Pre-process the data before alignment
+  size_t preProcElts = min(ptrAlign - data, size);
+  // Read memory by T (slow). One iter is enough bc the number of threads would
+  // be bigger than `preProcElts`
+  if (offset < preProcElts) {
+    if (isnan(data[offset])) __trap();
+  }
+  // We have processes this amount of data
+  size -= preProcElts;
+
+  // Start BytePack processing
+  BytePack* ptr = (BytePack*)ptrAlign;
+  // Size of input data in unit of BytePack
+  size_t sizeInBP = size * sizeof(T) / sizeof(BytePack);
+  // Number of BytePacks processed in one fast-path iteration
+  size_t loopSize = blockDim.x * gridDim.x * UNROLL;
+
+  // Fast path
+  // The condition below makes sure there is enough data to process (`loopSize`)
+  for (; offset + loopSize <= sizeInBP; offset += loopSize) {
+    checkChunk<T>(ptr + offset);
+  }
 
-  for (size_t i = tid; i < size; i += stride) {
-    CUDA_KERNEL_ASSERT(!isnan(data[i]));
+  // The rest data goes on slow path
+  // We just do regular load and check
+  for (; offset < sizeInBP; offset += blockDim.x * gridDim.x) {
+    BytePack tmp = ptr[offset];
+    CheckBytePack<T, sizeof(BytePack)/sizeof(T)>::check(&tmp);
```

Comment:
confused, if we are sure we have enough data left for one call to CheckBytePack<T, B/T> doesn't that also mean we have enough data for a faster call to CheckBytePack<T>?

Reply:
Do you mean why not a faster call to **checkChunk**? 
The reason is that `checkChunk` checks on `8*BytePack` in one call.
This slow loop here accounts for the case that we don't have `8*BytePack` left.

Reply:
oh- yes, i got confused between the two.  i think this makes sense.

Reply:
So in summary, i think the algorithm is

Pre: always `< 1` BytePack, since `=1`  would imply its already 16B aligned, so don't even use 'CheckByte', do a local check

Body: process chunks of 8 BytePack (e.g. 8*16 =  128B chunks) per call

Tail: since alignment is now guaranteed, just process (N < 8) 16B Bytepacks individually